File: kaiix/tree/BalancedBinaryTree.java
package tree;

/*
 * http://oj.leetcode.com/problems/balanced-binary-tree/
 */
public class BalancedBinaryTree {
    public static class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode(int x) { val = x; }
    }

    public static boolean isBalanced(TreeNode root) {
        if (root != null) {
            if (Math.abs(heightOfTree(root.left) - heightOfTree(root.right)) <= 1) {
                return isBalanced(root.left) && isBalanced(root.right);
            } else {
                return false;
            }
        } else {
            return true;
        }
    }

    private static int heightOfTree(TreeNode node) {
        if (node == null) {
            return 0;
        } else {
            return 1 + Math.max(heightOfTree(node.left), heightOfTree(node.right));
        }
    }

    public static void main(String []args){
        TreeNode root = new TreeNode(1);
        root.left = new TreeNode(2);
        root.left.left = new TreeNode(3);
        root.left.left.left = new TreeNode(4);
        root.right = new TreeNode(2);
        root.right.right = new TreeNode(3);

        System.out.println("balanced = " + isBalanced(root));
    }
}

File: dave/tree/BalancedBinaryTree.py
# Definition for a  binary tree node
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


class Solution:
    # @param root, a tree node
    # @return a boolean
    def isBalanced(self, root):
        return self.balancedHeight(root) >= 0

    def balancedHeight(self, root):
        if not root:
            return 0
        left = self.balancedHeight(root.left)
        right = self.balancedHeight(root.right)
        if left < 0 or right < 0:
            return -1
        if left - right > 1 or right - left > 1:
            return -1
        return max(left, right) + 1
